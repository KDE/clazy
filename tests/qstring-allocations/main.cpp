#include <QtCore/QString>
#include <QtCore/QStringList>
#include <QtCore/QDebug>

const char * some_char_pointer_indirection(const char *)
{
    return nullptr;
}

const char * some_char_pointer()
{
    return nullptr;
}

void test()
{
    QString s1;
    const char *cstring = "foo";
    s1.contains("string"); // Warning
    s1.contains(some_char_pointer()); // OK

    QString s2 = "string"; // Warning
    QString s3 = QStringLiteral("string"); // OK
    QString s4(some_char_pointer()); // OK

    s1 += "foo"; // Warning
    s1 += QStringLiteral("foo"); // OK
    s1 += QLatin1String("foo"); // OK

    QString s5 = QString::fromLatin1("BAR"); // Warning
    QString s6 = QString::fromUtf8("BAR"); // Warning
    QString s7 = QString::fromLatin1(some_char_pointer()); // OK
    QString s8 = QString::fromUtf8(some_char_pointer()); // OK
    QString s81 = QString::fromUtf8(some_char_pointer_indirection("foo")); // OK
    QString s123 = QString::fromLatin1(some_char_pointer_indirection("foo")); // OK

    QString s9 = QLatin1String("string"); // Warning
    s9 =  QLatin1String("string"); // Warning
    s9 =  QLatin1String(some_char_pointer_indirection("foo")); // OK
    QString s10 = true ? QLatin1String("string1") : QLatin1String("string2"); // Warning
    s10 = true ? QLatin1String("string1") :  QLatin1String("string2"); // Warning

    QString s11 = QString(QLatin1String("foo")); // Warning
    QStringList stringList;
    stringList << QString::fromLatin1("foo", 1); // OK
    QString s12 = QLatin1String(""); // OK, QString is optimized for the empty case
    QString s = QLatin1String(cstring + sizeof("foo")); // OK
    s = QString::fromLatin1("bar %1").arg(1); // Warning
    s = QString::fromLatin1(true ? "foo" : "foo2").arg(1); // Warning
    s += QString::fromLatin1(  // Warning: Multi-line, to test FIXIT
        "foo" );
    QString::fromLatin1(true ? cstring : "cstring"); // OK
    QString escaped = "\\\\"; // Warning
    QString s13 = QString("foo"); // Warning
    QString s14 = QString(""); // Warning
    QString s15 = ""; // Warning
    s15.startsWith(QLatin1String("ff")); // OK
    s15.startsWith("ff"); // Warning, should be QLatin1String
    s15.contains("ff"); // Warning, should be QStringLiteral
    s15.indexOf("ff"); // Warning, should be QStringLiteral
}








struct A
{
    A() : member(QLatin1String("foo")) {} // Warning
    QString member;
};

struct A2
{
    A2() : member(QStringLiteral("foo")) // OK
         , member2(QString(QLatin1String("foo"))) {} // Warning
    void  test(const QString &calendarType = QLatin1String("gregorian") ); // Warning
    QString member;
    QString member2;
};

void test3()
{
    QString s;
    if (s == QString::fromLatin1("foo")) // Warning
        return;
    if (s == QString::fromLatin1("foo %1").arg(1)) // Warning
        return;
}

void test4(const QString &) {}

void test5()
{
    test4(QString::fromLatin1("foo")); // Warning
    QString s;
    s.contains(QString::fromLatin1("a")); // Warning
    s.compare(QStringLiteral("a"), QString::fromLatin1("a")); // Warning
}

#include <exception>
void exceptionTest() // Just to test a crash
{
    try {

    } catch (std::exception e) {

    }
}

QStringList returns_stringlist(QString) { return {}; }

void test_foreach()
{
    foreach (const QString &s, returns_stringlist(QLatin1String("bar"))) // Warning
    {
    }
}

void testCharsets()
{
    QString s = "é";
    s += "é";
    s += "e";
    s.compare("é", "e");
    s.startsWith("é");
    s.startsWith("e");
    foreach (const QString &s, returns_stringlist(QLatin1String("éÈ"))) { }
    QString s11 = QString(QLatin1String("é"));

    if (s == QString::fromLatin1("foó"))
        return;
    if (s == QString::fromLatin1("foó %1").arg(1))
        return;
    s.contains(QString::fromLatin1("é"));
    s += QString::fromLatin1(
        "fóó" );
    QString s123 = QString::fromLatin1(some_char_pointer_indirection("fóó"));
    QString s9 = QLatin1String("stringéééé");
}

struct TestCharsets
{
    TestCharsets() : member(QStringLiteral("é"))
         , member2(QString(QLatin1String("é"))) {} // Warning
    void  test(const QString &calendarType = QLatin1String("éééé´e") );
    void test2();
    QString member;
    QString member2;
};











void moreShouldBeQLatin1String()
{
    QString s;
    bool b = (s == QString::fromLatin1("="));
    if (s.startsWith("Version=5")) {}
    if (s.startsWith(QLatin1String("Version=5"))) {} // OK
    QString result = QString::fromLatin1("Enum") + s;
}

void TestCharsets::test2()
{
    QString s = QString::fromLatin1("é"); // Can't fix this weird one
    QString s2 = QString::fromUtf8("é");
    s2 += QString::fromLatin1("é");
    s2 += QString::fromUtf8("é");
    s = "á";
    s += "á";
    QString("á");
    QStringList slist;
    slist << "á" << "a";
}





#include <QtCore/QRegExp>
void testBlacklistedQRegExp()
{
    QRegExp exp1("literal");
    QRegExp exp2(QLatin1String("literal"));
}


void charsetEdgeCase()
{
    test4(QLatin1String("ááá"));
}

void testQDebug()
{
    qDebug() << "test"; // OK
}

void testEmpty()
{
    QStringList list = QStringList() << "foo" << "";
}


void test_bug358732()
{
    QString("foo").sprintf("0x%02X", 0x1E); // Warn and use QSL
    QString("").sprintf("0x%02X", 0x1E); // Warn and use QSL
}


struct QTestData {};

template<typename T>
QTestData &operator<<(QTestData &data, const T &value);

void test_QTestData()
{
    QTestData t;
    t << QString::fromLatin1("foo");
}

void test_QStringList_streamOp()
{
    QStringList list;
    list << QStringLiteral("foo") << QLatin1String("foo") << QString::fromLatin1("foo") << "foo";
    list += QStringLiteral("foo");
    list += QLatin1String("foo");
    list += QString::fromLatin1("foo");
    list += "foo";
}

void testQLatin1String2Args()
{
    QString s2 = QLatin1String("foo", 3);
}

QString s = QString::fromUtf8("ö");
QString s2 = QString::fromUtf8("\xc3\xb6");
QString s3 = true ? "ö" : "\xc3\xb6";
