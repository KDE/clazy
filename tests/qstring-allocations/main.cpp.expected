qstring-allocations/main.cpp:19:17: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:22:18: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:26:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:30:18: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:31:18: warning: QString::fromUtf8() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:37:18: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:38:5: warning: QString::operator=(QLatin1String("literal") [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:40:19: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:41:5: warning: QString::operator=(QLatin1String("literal") [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:43:19: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:48:9: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:49:9: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:50:10: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:53:23: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:54:19: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:55:19: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:56:19: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:58:20: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:59:18: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:60:17: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:72:11: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:79:20: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:80:46: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:88:14: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:90:14: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:98:11: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:100:16: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:101:36: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:118:51: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:118:51: warning: FixIt failed, requires manual intervention:  Can't use QStringLiteral in macro [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:125:17: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:126:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:127:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:128:15: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:128:21: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:129:18: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:130:18: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:131:51: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:131:51: warning: FixIt failed, requires manual intervention:  Can't use QStringLiteral in macro [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:132:41: warning: Don't use QLatin1String with non-latin1 literals [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:132:19: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:134:14: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:136:14: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:138:16: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:139:10: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:142:32: warning: Don't use QLatin1String with non-latin1 literals [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:142:18: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:148:42: warning: Don't use QLatin1String with non-latin1 literals [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:148:20: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:149:60: warning: Don't use QLatin1String with non-latin1 literals [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:149:46: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:168:20: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:169:22: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:171:22: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:176:17: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:177:18: warning: QString::fromUtf8() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:178:11: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:179:11: warning: QString::fromUtf8() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:180:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:181:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:182:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:184:14: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:184:22: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:201:25: warning: Don't use QLatin1String with non-latin1 literals [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:201:11: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:211:41: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:211:50: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:217:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:218:5: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:230:10: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:236:38: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:236:62: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:236:92: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:238:13: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:239:13: warning: QString::fromLatin1() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:240:13: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:245:18: warning: QString(QLatin1String) being called [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:248:13: warning: QString::fromUtf8() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:249:14: warning: QString::fromUtf8() being passed a literal [-Wclazy-qstring-allocations]
qstring-allocations/main.cpp:250:14: warning: QString(const char*) being called [-Wclazy-qstring-allocations]
